File: src/main/java/com/puhovin/intellijplugin/twm/action/AbstractApplyAvailabilitiesAction.java
package com.puhovin.intellijplugin.twm.action;

import com.intellij.openapi.actionSystem.AnAction;
import com.intellij.openapi.actionSystem.AnActionEvent;
import com.intellij.openapi.project.Project;
import com.puhovin.intellijplugin.twm.core.ToolWindowManagerDispatcher;
import com.puhovin.intellijplugin.twm.model.ToolWindowPreference;
import org.jetbrains.annotations.NotNull;

import java.util.HashMap;
import java.util.List;
import java.util.Map;

public abstract class AbstractApplyAvailabilitiesAction extends AnAction {

    @Override
    public void actionPerformed(AnActionEvent e) {
        Project project = e.getProject();
        if (project == null) return;

        ToolWindowManagerDispatcher dispatcher = ToolWindowManagerDispatcher.getInstance(project);

        List<ToolWindowPreference> prefs = getPreferencesToApply(dispatcher);
        Map<String, ToolWindowPreference> newPrefs = new HashMap<>();
        prefs.forEach(pref -> newPrefs.put(pref.getId(), pref));

        dispatcher.applyPreferences(newPrefs);
    }

    @NotNull
    protected abstract List<ToolWindowPreference> getPreferencesToApply(@NotNull ToolWindowManagerDispatcher dispatcher);
}

File: src/main/kotlin/com/puhovin/intellijplugin/twm/action/AbstractApplyAvailabilitiesAction.kt
package com.puhovin.intellijplugin.twm.action

import com.intellij.openapi.actionSystem.AnAction
import com.intellij.openapi.actionSystem.AnActionEvent
import com.puhovin.intellijplugin.twm.core.ToolWindowManagerDispatcher
import com.puhovin.intellijplugin.twm.model.ToolWindowPreference

/**
 * Abstract base action for applying tool window preferences.
 */
abstract class AbstractApplyAvailabilitiesAction : AnAction() {

    /**
     * Executes the action to apply the tool window preferences.
     *
     * @param e The action event
     */
    override fun actionPerformed(e: AnActionEvent) {
        val project = e.project ?: return

        val dispatcher = ToolWindowManagerDispatcher.getInstance(project)
        dispatcher.applyPreferences(getPreferencesToApply(dispatcher))
    }

    /**
     * Abstract method to get a list of tool window preferences to apply.
     *
     * @param dispatcher The tool window manager dispatcher
     * @return A list of preferences to apply
     */
    protected abstract fun getPreferencesToApply(dispatcher: ToolWindowManagerDispatcher): List<ToolWindowPreference>
}